test/controllers/api_users/permissions_controller_test.rb

```diff
@@ -0,0 +1,347 @@
+      create(:access_token, application:, resource_owner_id: api_user.id)
+
+      current_user = create(:superadmin_user)
+      sign_in current_user
+
+      stub_policy current_user, api_user, update?: true
+
+      other_permission = other_application.supported_permissions.find_by(name: "other")
+
+      patch :update, params: { api_user_id: api_user, application_id: application, application: { supported_permission_ids: [other_permission.id] } }
+
+      api_user.reload
+
+      assert_equal [], api_user.supported_permissions
+    end
+
+    should "prevent permissions being added that are not grantable from the ui" do
+      application = create(:application, with_supported_permissions: %w[other], with_supported_permissions_not_grantable_from_ui: %w[not_from_ui])
+      api_user = create(:api_user)
+      create(:access_token, application:, resource_owner_id: api_user.id)
+
+      current_user = create(:superadmin_user)
+      sign_in current_user
+
+      stub_policy current_user, api_user, update?: true
+
+      not_from_ui_permission = application.supported_permissions.find_by(name: "not_from_ui")
+
+      patch :update, params: { api_user_id: api_user, application_id: application, application: { supported_permission_ids: [not_from_ui_permission.id] } }
+
+      api_user.reload
+
+      assert_equal [], api_user.supported_permissions
+    end
+
+    should "assign the application id to the application_id flash" do
+      application = create(:application, with_supported_permissions: %w[permission])
+      api_user = create(:api_user)
+      create(:access_token, application:, resource_owner_id: api_user.id)
+
+      current_user = create(:superadmin_user)
+      sign_in current_user
+
+      stub_policy current_user, api_user, update?: true
+
+      permission = application.supported_permissions.find_by(name: "permission")
+
+      patch :update, params: { api_user_id: api_user, application_id: application, application: { supported_permission_ids: [permission.id] } }
+
+      assert_equal application.id, flash[:application_id]
+    end
+
+    should "raise an exception if the user cannot be found" do
+      application = create(:application)
+
+      current_user = create(:superadmin_user)
+      sign_in current_user
+
+      assert_raises(ActiveRecord::RecordNotFound) do
+        patch :update, params: { api_user_id: "unknown-id", application_id: application, application: { supported_permission_ids: %w[id] } }
+      end
+    end
+
+    should "prevent unauthorised users" do
+      application = create(:application)
+      api_user = create(:api_user)
+      create(:access_token, application:, resource_owner_id: api_user.id)
+
+      current_user = create(:superadmin_user)
+      sign_in current_user
+
+      stub_policy current_user, api_user, update?: false
+
+      patch :update, params: { api_user_id: api_user, application_id: application, application: { supported_permission_ids: [] } }
+
+      assert_not_authorised
+    end
```

Comment:
Minor: It might be nice to move this test nearer the top of this context under the `should "prevent unauthenticated users"` test like in the context for the `#edit` action. At first I thought you'd missed out a test like this, because it wasn't where I expected it.

Reply:
We copied these tests from the equivalent users tests so inherited the problem from there. We're not going to move the test for now but think we might tackle this as part of trying to remove duplication between the 3 permissions controllers we've ended up with.